EZP-26057: Permissions API

pspanja opened the pull request:
This PR resolves https://jira.ez.no/browse/EZP-26057

As a preparation for https://jira.ez.no/browse/EZP-26000, this implements permission API.
Methods are deprecated in `Repository` and implemented in new PermissionResolver:

- `getCurrentUserReference()`
- `setCurrentUserReference()`
- `hasAccess()`
- `canUser()`

The implementation in `Repository` now falls back to the `PermissionResolver`.
Internal method `sudo()` is also implemented in the new `PermissionResolver`.
Protected properties in the Core implementation of `Repository` are also marked as deprecated:

- `$currentUser`
- `$currentUserRef`

## TODOs
- [x] Add integration tests (copy from Repository)
- [x] Undeprecate `Repository::sudo()`, mark `PermissionResolver::sudo()` internal and make second argument mandatory so dependency on repository in constructor can be removed.

pspanja commented:
Review ping @andrerom @alongosz @bdunogier

andrerom commented:
> Review ping andrerom alongosz bdunogier 

bdunogier is out catching pokemons this week so you don't need to ping him :)

andrerom commented on eZ/Publish/API/Repository/PermissionService.php:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $user
+     */
+    public function setCurrentUser(UserReference $user);
```
while at it `setUserReference`

pspanja replied:
You mean adding that method? It's already possible with this one, as `User` implements `UserReference`.

BTW, I was thinking about adding `newUserReference($id)`, so it is not necessary to instantiate the class from `Core` namespace.

andrerom replied:
> You mean adding that method? It's already possible with this one, as User implements UserReference.

No, I was just referring to cleaning up the naming.

pspanja replied:
+1 indeed, will update.

pspanja replied:
Renamed to `setCurrentUserReference` in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/8a67e929c7792499b51b5a33c957233fbbdad124.

andrerom commented on eZ/Publish/Core/Repository/Repository.php:
```diff
@@ -879,6 +744,28 @@ public function getFieldTypeService()
+    {
+        if ($this->permissionService === null) {
+            $this->permissionService = new PermissionService(
+                $this,
```
should avoid this if possible

pspanja replied:
We need it only for `sudo()`, should be possible to get rid of that once we restructure repo implementation (introduce domain logic implementation).

andrerom commented on eZ/Publish/Core/Repository/Repository.php:
```diff
@@ -879,6 +744,28 @@ public function getFieldTypeService()
+        if ($this->permissionService === null) {
+            $this->permissionService = new PermissionService(
+                $this,
+                $this->getUserService(),
```
and this

pspanja replied:
We need `UserDomainMapper` for that, can be separate issue.

andrerom replied:
You are referring to for `getCurrentUser` right?
I'm somewhat leaning toward omitting that and only having `getCurrentUserRef` as this service does not really care about the full user.

This implies we should cover this use case somewhere higher up before we remove the deprecated `Repository->getCurrentuser()`.

pspanja replied:
That makes sense to me, I'll remove it.

pspanja replied:
Removed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/53be41ef3fb217330a7127d15551c26989f711d2.

andrerom commented on eZ/Publish/Core/Repository/Repository.php:
```diff
@@ -879,6 +744,28 @@ public function getFieldTypeService()
+                $this->getRoleDomainMapper(),
+                $this->getLimitationService(),
+                $this->persistenceHandler->userHandler(),
+                $this->currentUserRef,
```
and this

pspanja replied:
Not sure we can do that ATM, as it is injected in the `Repository`.

andrerom commented on eZ/Publish/Core/Repository/Repository.php:
```diff
@@ -879,6 +744,28 @@ public function getFieldTypeService()
+                $this->getLimitationService(),
+                $this->persistenceHandler->userHandler(),
+                $this->currentUserRef,
+                $this->currentUser
```
and this

pspanja replied:
I use this from `Repository`, so that protected properties there can be deprecated.
Otherwise we could move the logic from the constructor there to the PermissionService.

crevillo commented on eZ/Publish/API/Repository/PermissionService.php:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * @param string $module The module, aka controller identifier to check permissions on
+     * @param string $function The function, aka the controller action to check permissions on
+     * @param \eZ\Publish\API\Repository\Values\User\UserReference $user
```
nitpick: if user can be null i'd add that in the doc too.

andrerom replied:
no it can't, it will be anonymous user if nothing else is set.

pspanja replied:
c/p from `Repository`, will update.

pspanja replied:
Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/b4c4758c7f1a4718d547eb18277d550d69e6b462, also added some description while at it.

pspanja commented:
Continued from a now hidden thread.

>> BTW, I was thinking about adding newUserReference($id), so it is not necessary to manually instantiate the class from Core namespace.
>
> mm, let me think about that. id being the identifier of a user reference can be seen as implementation detail.
>
> So flow oop wise should rather then be something ala:
> newUserReference(User $user)
>
> Which is already possible indirectly using setCurrentUser -> getCurrentUserReference

Yeah, but you would need a full `User`, which defeats the purpose.

The motive is to be able to move repo implementation without causing breaks (in theory :)
Maybe we should have had `UserReference` implementation in API.

andrerom commented on eZ/Publish/Core/Repository/PermissionService.php:
```diff
@@ -0,0 +1,279 @@
+     *
+     * @return mixed
+     */
+    public function sudo(\Closure $callback, RepositoryInterface $outerRepository = null)
```
I was about to say leave sudo on Repository, it will be obsolete once we have have seperate permission layer.

However we need to keep tracking od sudo here, so maybe rather do this:
- mark as internal
- make outerRepository mandatory so we don't need to inject repository in constructor _(circular referance)_
- optionally: don't deprecate sudo on Repo just yet

pspanja replied:
Agreed, todo added. I'll undeprecate `Repository::sudo()` as we don't provide an alternative yet.

pspanja replied:
Updated in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/bffc066c1c0e679a97347739b506965fbafc789f (`Repository::sudo()` was actually not deprecated)

pspanja commented:
Added missing integration tests in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/c06e55502b26ddca4a2577b82f7f6622c6d3a7e7.

andrerom commented on eZ/Publish/API/Repository/PermissionService.php:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return bool
+     */
+    public function canUser($module, $function, ValueObject $object, $targets = null);
```
targets can be array here in the new api, it's not type hinted in repo since it was single value and was changed to multi value by you in 2a68f9e67cbec67c9561b137d1fffa19ddcade76 ;)

pspanja replied:
Changed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/08bd32198e7a555ad859f98533da93f39bba3695. Note that input validation is now moved from `PermissionService` to `Repository`, where `null` is converted to empty array before passing to `PermissionService::canUser()`. There we still need to convert it back to `null`, because that is what `LimitationType::evaluate()` expects.

We should update that as well with next major version bump.

andrerom commented:
besides api nitpick I think this is ready for final review, right?

pspanja commented:
Yes indeed, fire away (will take care of type hint)

andrerom commented:
review ping @alongosz

andrerom commented:
Afaik we can skip the single slot implementation and just pass the underlying instance right?

pspanja commented:
> Afaik we can skip the single slot implementation and just pass the underlying instance right?
just adds call overhead it seems, and we won't add operations to.

Yes, and the same could be done with `FieldTypeService`, though I guess we better wait with that one.

> A bit late, but if you btw have suggestion for better name then "service" to have distinction between domain services and helper something services like this., I'm all ears :)

Maybe `PermissionResolver`?

andrerom commented:
> Yes, and the same could be done with FieldTypeService, though I guess we better wait with that one.

separate PR for FieldTypeService, but please do and no issue needed if it is just returning the underlying one.

> Maybe PermissionResolver?

sounds good

pspanja commented:
Removed `SignalSlot` implementation in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/95ed8779dacf11f4e2189d725a84b9fd3f60b60d
Renamed to `PermissionResoolver` in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/7dbbb7c7caaf727c8e41da1232fec7b846c94cf4

Also rebased on master, now ready for final review: ping @andrerom @bdunogier @alongosz

andrerom commented:
+1

andrerom commented on eZ/Publish/Core/Repository/PermissionResolver.php:
```diff
@@ -0,0 +1,266 @@
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\Core\Repository;
```
Note: I'd like this to be in `eZ\Publish\Core\Repository\Permission;` so we already start to aim for the separation in `Core\Repository` between "Site" => SingleSlot => Permissions => Cache => "BusinessLogic".
So we don't break class naming of more then we need.

pspanja replied:
It does, I'll go for `eZ\Publish\Core\Repository\Permission` (singular), like in https://github.com/ezsystems/ezpublish-kernel/pull/1714 if you agree.

alongosz commented on eZ/Publish/API/Repository/Repository.php:
```diff
@@ -19,6 +19,8 @@
 interface Repository
 {
     /**
+     * @deprecated since 6.5, to be removed. Use PermissionResolver::getCurrentUserReference() instead.
```
I might be wrong here, but I think we should deprecate since `6.6` because `6.5` has been released already.

pspanja replied:
Indeed, changing.

pspanja replied:
Updated in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/b36e667862234896be49c144d5d43b82dbb7d9b7

pspanja commented:
Moved `PermissionResolver` implementation to `Core\Permission` in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/2a9ed92a1ab0ed2b5edb76a5a3ec73ab09a0b0f7

alongosz commented on eZ/Publish/Core/Repository/PermissionsCriterionHandler.php:
```diff
@@ -23,13 +24,27 @@
+        PermissionResolverInterface $permissionResolver,
+        LimitationService $limitationService
+    ) {
+        $this->PermissionResolver = $permissionResolver;
```
it should be `$this->permissionResolver = $permissionResolver;`

pspanja replied:
Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/c3a60b9d14944b963ca3998021c5880dfba328bc

alongosz commented on eZ/Publish/Core/Repository/PermissionsCriterionHandler.php:
```diff
@@ -77,7 +92,7 @@ public function addPermissionsCriterion(Criterion &$criterion)
     public function getPermissionsCriterion($module = 'content', $function = 'read')
     {
-        $permissionSets = $this->repository->hasAccess($module, $function);
+        $permissionSets = $this->PermissionResolver->hasAccess($module, $function);
```
same as mentioned above, `$this->permissionResolver`

pspanja replied:
Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/c3a60b9d14944b963ca3998021c5880dfba328bc

alongosz commented on eZ/Publish/Core/Repository/PermissionsCriterionHandler.php:
```diff
@@ -92,9 +107,8 @@ public function getPermissionsCriterion($module = 'content', $function = 'read')
          * role limitation, otherwise it will be merged into RoleAssignment's OR condition.
          */
-        $currentUserRef = $this->repository->getCurrentUserReference();
+        $currentUserRef = $this->PermissionResolver->getCurrentUserReference();
```
same as mentioned above, `$this->permissionResolver`

pspanja replied:
Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/c3a60b9d14944b963ca3998021c5880dfba328bc

andrerom commented:
+1

alongosz commented on eZ/Publish/Core/Repository/Tests/Service/Mock/PermissionsCriterionHandlerTest.php:
```diff
@@ -20,19 +19,23 @@
-            $repositoryMock,
-            'repository',
+            $permissionResolverMock,
+            'PermissionResolver',
```
@pspanja: Travis fails here - also should be `permissionResolver`. I missed that when looking for occurences of this.

pspanja replied:
Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/1720/commits/8ec6563592431ce8be88b138ffc1d2e45ec3e263

alongosz commented:
+1 after a fix mentioned in a diff above ;)

pspanja commented:
Merging, thanks all for the review.
